include/material/norsand.tcc

```diff
@@ -0,0 +1,627 @@
+                                                 FailureState yield_type) {
+
+  // Get state variables
+  const double mean_p = (*state_vars).at("p");
+  const double deviatoric_q = (*state_vars).at("q");
+  const double M_theta = (*state_vars).at("M_theta");
+  const double p_cohesion = (*state_vars).at("p_cohesion");
+  const double p_dilation = (*state_vars).at("p_dilation");
+  double p_image;
+  double e_image;
+
+  if (yield_type == mpm::NorSand<Tdim>::FailureState::Elastic) {
+    // Keep the same pressure image and void ratio image at critical state
+    p_image = (*state_vars).at("p_image");
+    e_image = (*state_vars).at("e_image");
+  } else {
+    // Compute and update pressure image
+    p_image =
+        (mean_p + p_cohesion) *
+            std::pow(
+                ((1 - N_ / M_theta * deviatoric_q / (mean_p + p_cohesion)) /
+                 (1 - N_)),
+                ((N_ - 1) / N_)) -
+        p_cohesion - p_dilation;
+    (*state_vars).at("p_image") = p_image;
+
+    // Compute and update void ratio image
+    // e_image = e_max_ - (e_max_ - e_min_) / log(crushing_pressure_ / p_image);
+    e_image = gamma_ - lambda_ * log(p_image / reference_pressure_);
+
+    if (e_image < 1.0E-15) e_image = 1.0E-15;
+
+    (*state_vars).at("e_image") = e_image;
+  }
+
+  // Update void ratio
+  // Note that dstrain is in tension positive - depsv = de / (1 + e_initial)
+  double dvolumetric_strain = dstrain(0) + dstrain(1) + dstrain(2);
+  double void_ratio = (*state_vars).at("void_ratio") -
+                      (1 + void_ratio_initial_) * dvolumetric_strain;
+  if (void_ratio < 1.0E-15) void_ratio = 1.0E-15;
+  (*state_vars).at("void_ratio") = void_ratio;
+
+  // Compute and update psi
+  double psi_image = void_ratio - e_image;
+  (*state_vars).at("psi_image") = psi_image;
+
+  return true;
+}
+
+//! Compute elastic tensor
+template <unsigned Tdim>
+bool mpm::NorSand<Tdim>::compute_p_bond(mpm::dense_map* state_vars) {
+
+  // Compute current zeta cohesion
+  double zeta_cohesion = exp(-m_cohesion_ * (*state_vars).at("epds"));
+
+  if (zeta_cohesion > 1.)
+    (*state_vars).at("zeta_cohesion") = 1.;
+  else if (zeta_cohesion < 1.0E-15)
+    (*state_vars).at("zeta_cohesion") = 0.;
+  else
+    (*state_vars).at("zeta_cohesion") = zeta_cohesion;
+
+  // Update p_cohesion
+  double p_cohesion = p_cohesion_initial_ * (*state_vars).at("zeta_cohesion");
+  (*state_vars).at("p_cohesion") = p_cohesion;
+
+  // Compute current zeta dilation
+  double zeta_dilation = exp(-m_dilation_ * (*state_vars).at("epds"));
+
+  if (zeta_dilation > 1.)
+    (*state_vars).at("zeta_dilation") = 1.;
+  else if (zeta_dilation < 1.0E-15)
+    (*state_vars).at("zeta_dilation") = 0.;
+  else
+    (*state_vars).at("zeta_dilation") = zeta_dilation;
+
+  // Update p_dilation
+  double p_dilation = p_dilation_initial_ * (*state_vars).at("zeta_dilation");
+  (*state_vars).at("p_dilation") = p_dilation;
+
+  return true;
```

Comment:
Could you explain where does this fail to return true? Why is this function defined as a bool?

Reply:
I will change this to void then